Dependencies
============

Tychus utilizes a number of open source projects to run. Please check out the following links for more information about each tool used in the Tychus pipeline.

* `Abyss <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2694472/>`_ *De novo* sequence assembler for short-paired reads
* `BCFtools <https://github.com/samtools/bcftools>`_ Variant call format (VCF) manipulation
* `Bowtie2 <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3322381/>`_ Short-read sequence aligner
* `CISA <http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0060843q>`_ Contig integrator
* `Docker <https://www.docker.com/what-docker>`_ Software containerization platform
* `Figtree <http://tree.bio.ed.ac.uk/software/figtree/>`_ Phylogenetic tree viewer
* `Freebayes <https://arxiv.org/abs/1207.3907>`_ Haplotype-based variant caller
* `IDBA-UD <http://i.cs.hku.hk/~chin/paper/idba_ud-revised-latest.pdf>`_ Short-read genome assembler
* `KmerGenie <https://arxiv.org/pdf/1304.5665.pdf>`_ Optimal kmer selection for building De-Bruijn graphs
* `kSNP3 <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3857212/>`_ SNP discovery and phylogeny
* `Nextflow <https://www.nextflow.io>`_ Workflow framework
* `Prokka <https://www.ncbi.nlm.nih.gov/pubmed/24642063>`_ Prokaryotic genome annotation tool
* `QUAST <https://github.com/ablab/quast>`_ Quality assessment tool for genome assemblies
* `Samtools <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/>`_ SAM/BAM alignment processor
* `SPades <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3342519/>`_ Assembler for single-celled bacterial genomes
* `Trimmomatic <http://bioinformatics.oxfordjournals.org/content/early/2014/04/01/bioinformatics.btu170>`_ Read trimmer and quality control
* `Velvet <http://genome.cshlp.org/content/genome/18/5/821.full.html>`_ *De novo* short-read assembler
